Content.Shared/Charges/Systems/SharedChargesSystem.cs

```diff
@@ -0,0 +1,59 @@
+using Content.Shared.Charges.Components;
+using Content.Shared.Examine;
+
+namespace Content.Shared.Charges.Systems;
+
+public abstract class SharedChargesSystem : EntitySystem
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<LimitedChargesComponent, ExaminedEvent>(OnExamine);
+    }
+
+    protected virtual void OnExamine(EntityUid uid, LimitedChargesComponent comp, ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        args.PushMarkup(Loc.GetString("limited-charges-charges-remaining", ("charges", comp.Charges)));
+        if (comp.Charges == comp.MaxCharges)
+        {
+            args.PushMarkup(Loc.GetString("limited-charges-max-charges"));
+            return;
+        }
+    }
+
+    /// <summary>
+    /// Tries to add a number of charges. If it over or underflows it will be clamped, wasting the extra charges.
+    /// </summary>
+    public void AddCharges(LimitedChargesComponent comp, int change)
+    {
+        var old = comp.Charges;
+        comp.Charges = Math.Clamp(comp.Charges + change, 0, comp.MaxCharges);
+        if (comp.Charges != old)
+            Dirty(comp);
+    }
+
+    /// <summary>
+    /// Gets the limited charges component and returns true if there are no charges. Will return false if there is no limited charges component.
+    /// </summary>
+    public bool IsEmpty(EntityUid uid, out LimitedChargesComponent? comp)
```

Comment:
you can save yourself a null check after using this by adding the `[NotNullWhen(true)]` attribute to the out parameter here

Reply:
actually yeah, that’s true: i guess it depends on if the component passed out is anticipated to be used if the charge count is empty

if you do return before any code path that uses that parameter is executed, though, then it makes sense, otherwise it might be a bit messy 

on the other hand though, you could use the `Resolve` pattern with AddCharges and pass in the component ref there, null or not, but the way this is structured makes it seem like you’re trying to make that function dependent on the usage of this one

Reply:
i think it would be nicer to adhere to the pattern that other systems use where if you're attempting to use a function in a system that deals with an entity, it tries to resolve the components it needs to, i.e.,

```cs
public bool IsEmpty(EntityUid uid, LimitedChargesComponent? comp = null)
{
    if (!Resolve(uid, ref comp))
    {
        return false;
    }
    // etc...
}
```

and the same would apply to AddCharges as well